discovery-provider/src/utils/ipfs_lib.py

```diff
@@ -25,48 +25,67 @@ def __init__(self, ipfs_peer_host, ipfs_peer_port):
+    def get_metadata(self, multihash, default_metadata_fields, user_replica_set=None):
+        """Retrieve file from IPFS or gateway, validating metadata requirements prior to
         returning an object with no missing entries
         """
         logger.warning(f"IPFSCLIENT | get_metadata - {multihash}")
-        api_metadata = metadata_format
-        retrieved_from_local_node = False
+        api_metadata = default_metadata_fields
         retrieved_from_gateway = False
+        retrieved_from_local_node = False
         start_time = time.time()
 
-        # First try to retrieve from local ipfs node.
-        try:
-            api_metadata = self.get_metadata_from_ipfs_node(multihash, metadata_format)
-            retrieved_from_local_node = api_metadata != metadata_format
-        except Exception:
-            logger.error(
-                f"Failed to retrieve CID from local node, {multihash}", exc_info=True
-            )
-
-        # Else, try to retrieve from gateways.
-        if not retrieved_from_local_node:
-            try:
-                api_metadata = self.get_metadata_from_gateway(
-                    multihash, metadata_format, user_replica_set
-                )
-                retrieved_from_gateway = api_metadata != metadata_format
-            except Exception:
-                logger.error(
-                    f"Failed to retrieve CID from gateway, {multihash}", exc_info=True
-                )
+        with concurrent.futures.ThreadPoolExecutor(max_workers=2) as executor:
+            metadata_futures = {}
+            metadata_futures[executor.submit(
+                self.get_metadata_from_ipfs_node, multihash, default_metadata_fields
+            )] = 'metadata_from_ipfs_node'
+            metadata_futures[executor.submit(
+                self.get_metadata_from_gateway, multihash, default_metadata_fields, user_replica_set
+            )] = 'metadata_from_gateway'
+            for get_metadata_future in concurrent.futures.as_completed(
+                metadata_futures,
+                timeout=NEW_BLOCK_TIMEOUT_SECONDS
+            ):
+                original_task = metadata_futures[get_metadata_future]
+                try:
+                    api_metadata = get_metadata_future.result()
+                    retrieved = api_metadata != default_metadata_fields
+                    if retrieved:
+                        logger.info(
+                            f'IPFSCLIENT | retrieved metadata successfully, {api_metadata}, task: {original_task}'
+                        )
+                        if original_task == 'metadata_from_gateway':
+                            retrieved_from_gateway = True
+                        else:
+                            retrieved_from_local_node = True
+                        self.force_clear_queue_and_stop_task_execution(executor)
+                        break # use first returned result
+                except TimeoutError as e:
```

Comment:
do we need this TimeoutError exception? looks like we just raise it again, and it gets caught in the greater Exception block

Reply:
thanks!!!!!!! yes